From 07338e7dfd2253e87feadb8e271e19beee6e45ac Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Sat, 26 Jul 2025 22:58:10 +0200
Subject: [PATCH 6/7] net: pcs: airoha: add support for optional xfi reset line

On Airoha AN7583 there is a dedicated reset line for the PON XFI Serdes.
This is needed to permit changing the WAN sel register or the system
will stall on accessing the XFI register.

Add support for this optional dedicated reset to permit correct
configuration of the PON Serdes.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 drivers/net/pcs/airoha/pcs-airoha-common.c | 12 ++++++++++++
 drivers/net/pcs/airoha/pcs-airoha.h        |  1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/net/pcs/airoha/pcs-airoha-common.c b/drivers/net/pcs/airoha/pcs-airoha-common.c
index b4f84b85e7aa..8eaf5974084e 100644
--- a/drivers/net/pcs/airoha/pcs-airoha-common.c
+++ b/drivers/net/pcs/airoha/pcs-airoha-common.c
@@ -82,6 +82,10 @@ static int airoha_pcs_setup_scu(struct airoha_pcs_priv *priv,
 	const struct airoha_pcs_match_data *data = priv->data;
 	int ret;
 
+	ret = reset_control_assert(priv->xfi_rst);
+	if (ret)
+		return ret;
+
 	switch (data->port_type) {
 	case AIROHA_PCS_ETH:
 		airoha_pcs_setup_scu_eth(priv, interface);
@@ -91,6 +95,10 @@ static int airoha_pcs_setup_scu(struct airoha_pcs_priv *priv,
 		break;
 	}
 
+	ret = reset_control_deassert(priv->xfi_rst);
+	if (ret)
+		return ret;
+
 	/* TODO better handle reset from MAC */
 	ret = reset_control_bulk_assert(ARRAY_SIZE(priv->rsts),
 					priv->rsts);
@@ -989,6 +997,10 @@ static int airoha_pcs_probe(struct platform_device *pdev)
 	if (ret)
 		return dev_err_probe(dev, ret, "failed to get bulk reset lines\n");
 
+	priv->xfi_rst = devm_reset_control_get_optional_exclusive(dev, "xfi");
+	if (IS_ERR(priv->xfi_rst))
+		return dev_err_probe(dev, PTR_ERR(priv->xfi_rst), "failed to get xfi reset lines\n");
+
 	platform_set_drvdata(pdev, priv);
 
 	return 0;
diff --git a/drivers/net/pcs/airoha/pcs-airoha.h b/drivers/net/pcs/airoha/pcs-airoha.h
index 677ad58dd250..65cc535931f1 100644
--- a/drivers/net/pcs/airoha/pcs-airoha.h
+++ b/drivers/net/pcs/airoha/pcs-airoha.h
@@ -1173,6 +1173,7 @@ struct airoha_pcs_priv {
 	struct regmap *xfi_pma;
 	struct regmap *xfi_ana;
 
+	struct reset_control *xfi_rst;
 	struct reset_control_bulk_data rsts[AIROHA_PCS_MAX_NUM_RSTS];
 };
 
-- 
2.50.0

